dask/base.py

```diff
@@ -264,17 +265,19 @@ def visualize(*args, **kwargs):
 
 
 function_cache = {}
+function_cache_lock = threading.Lock()
 
 
 def normalize_function(func):
     try:
         return function_cache[func]
     except KeyError:
         result = _normalize_function(func)
-        if len(function_cache) >= 500:  # clear half of cache if full
-            for k in list(function_cache)[::2]:
-                del function_cache[k]
-        function_cache[func] = result
+        with function_cache_lock:
+            if len(function_cache) >= 500:  # clear half of cache if full
```

Comment:
Perhaps swap the `len` check with the lock? I'd expect the lock to only be needed on the clearing operation.